app/src/main/java/io/appium/uiautomator2/core/EventRegister.java

```diff
@@ -15,39 +16,54 @@
 
 public abstract class EventRegister {
 
+    public static int EVENT_COOLDOWN_MS = 750;
+
     public static Boolean runAndRegisterScrollEvents (ReturningRunnable<Boolean> runnable, long timeout) {
         // turn off listening to notifications since it interferes with us listening for the scroll
         // event here
         NotificationListener listener = NotificationListener.getInstance();
+
         boolean notificationListenerActive = listener.isListening;
         if (notificationListenerActive) {
             listener.stop();
         }
 
-        // say we want to listen to only scroll events
-        UiAutomation.AccessibilityEventFilter eventFilter = new UiAutomation.AccessibilityEventFilter() {
+        // here we set a callback for the accessibility event stream, keeping track of any scroll
+        // events we come across
+        AccessibilityEvent event = null;
+        final ArrayList<AccessibilityEvent> seenEvents = new ArrayList<>();
+        UiAutomation automation = UiAutomatorBridge.getInstance().getUiAutomation();
+        UiAutomation.OnAccessibilityEventListener onEvent = new UiAutomation.OnAccessibilityEventListener() {
             @Override
-            public boolean accept(AccessibilityEvent event) {
-                return event.getEventType() == AccessibilityEvent.TYPE_VIEW_SCROLLED;
+            public void onAccessibilityEvent(AccessibilityEvent accessibilityEvent) {
+                if (accessibilityEvent.getEventType() == AccessibilityEvent.TYPE_VIEW_SCROLLED) {
+                    seenEvents.add(accessibilityEvent);
+                }
             }
         };
+        automation.setOnAccessibilityEventListener(onEvent);
 
-        AccessibilityEvent event = null;
-        try {
-            event = UiAutomatorBridge.getInstance().getUiAutomation().executeAndWaitForEvent(runnable,
-                    eventFilter, timeout);
-            Logger.debug("Retrieved accessibility event for scroll");
-        } catch (TimeoutException ignore) {
-            Logger.error("Expected to receive a scroll accessibility event but hit the timeout instead");
+        // actually run the swipe/scroll
+        runnable.run();
+
+        // wait a certain amount of time for any accessibility events generated to be caught by
+        // our callback, then stop watching for events
+        try { Thread.sleep(timeout); } catch (InterruptedException ign) {}
```

Comment:
I really don't like this hard-coded sleep. What's wrong with `executeAndWaitForEvent`?
I can't see any possibility to get random event, because it [clear event queue before execute `Runnable`](
https://android.googlesource.com/platform/frameworks/base.git/+/master/core/java/android/app/UiAutomation.java#568).

Reply:
Here's what I observed with `executeAndWaitForEvent`: running the same full-screen swipe many many times, the swipe executed perfectly every time. However, the event retrieved by `executeAndWaitForEvent` had a different `scrollY` every time too. Sometimes it was very small, meaning the captured event happened at the very beginning of the swipe. In other words, I was not reliably getting the information I needed. This solution at least does a lot better job. I agree it's horribly hacky; maybe you can help me find a better way to be sure we're getting the last accessibility event and not one from mid-scroll.

Reply:
You can try it [this way ](https://android.googlesource.com/platform/frameworks/testing/+/master/uiautomator/library/core-src/com/android/uiautomator/core/InteractionController.java#360)